### PR TITLE
Request v2 consumers

### LINKED ISSUE #749: Request V2 Consumer

### PR DESCRIPTION
**Pull request by kevinhinterlong**
Resolves #749 

## Description
Adds consumers needed for request v2 interactions

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/core/addon/models/bard-request-v2/fragments/base.js [hunk 1]
```diff
@@ -54,4 +55,18 @@ export default class Base extends Fragment.extend(Validations) {
     }
     return this.metadataService.getById(this.type, this.field, this.source);
   }
+
+  @computed('field', 'parameters.{}')
+  get canonicalName() {
+    const { field: metric, parameters } = this;
+
+    return canonicalizeMetric({
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/fragments/base.js:63]: Are we using this `canonicalizeMetric` in a general way? Do we need to rename this?

File: packages/core/addon/models/bard-request-v2/request.js [hunk 2]
```diff
@@ -110,4 +119,280 @@ export default class Request extends Fragment.extend(Validations) {
       dataSource: clonedRequest.dataSource
     });
   }
+
+  /**
+   * @property {ColumnFragment} timeGrainColumn - The column containing the dateTime time-dimension
+   */
+  @computed('columns.[]')
+  get timeGrainColumn() {
+    return this.columns.find(column => column.type === 'time-dimension' && column.field === 'dateTime');
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/request.js:128]: Should this be 

`column.type === 'time-dimension'` OR `column.field === 'dateTime'`

Comment (jkusa) [reply packages/core/addon/models/bard-request-v2/request.js:128]: Should this be set of time-dimension?

File: packages/core/addon/models/bard-request-v2/request.js [hunk 3]
```diff
@@ -110,4 +119,280 @@ export default class Request extends Fragment.extend(Validations) {
...
+  }
+
+  /**
+   * @property {string} timeGrain - The grain parameter of the column containing the dateTime time-dimension
+   */
+  @computed('timeGrainColumn')
+  get timeGrain() {
+    return this.timeGrainColumn?.parameters?.grain;
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/request.js:136]: Do we want to find the finest time grain ?

File: packages/core/addon/models/bard-request-v2/request.js [hunk 4]
```diff
@@ -110,4 +119,280 @@ export default class Request extends Fragment.extend(Validations) {
...
+  @computed('filters.[]')
+  get dateTimeFilter() {
+    return this.filters.find(filter => filter.type === 'time-dimension' && filter.field === 'dateTime');
+  }
+
+  /**
+   * @property {Interval|undefined} interval - The interval to filter the dateTime for
+   */
+  @computed('dateTimeFilter.values')
+  get interval() {
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/request.js:151]: Are we assuming one time filter?

File: packages/core/addon/models/bard-request-v2/request.js [hunk 5]
```diff
@@ -110,4 +119,280 @@ export default class Request extends Fragment.extend(Validations) {
...
+  /**
+   * If feature flag enableRequestPreview disabled, then adds up to 1 of the column, otherwise adds a new column
+   * @param {ColumnMetadata} columnMetadataModel - the metadata for the column
+   * @param {object} parameters - the parameters to apply to the column
+   */
+  addColumnFromMeta(columnMetadataModel, parameters) {
+    const { columns } = this;
+    let columnExists = false;
+
+    if (!featureFlag('enableRequestPreview')) {
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/request.js:173]: Can we just assume request preview?

File: packages/core/addon/models/bard-request-v2/request.js [hunk 6]
```diff
@@ -110,4 +119,280 @@ export default class Request extends Fragment.extend(Validations) {
...
+    const { start, end } = newInterval.asStrings();
+    set(dateTimeFilter, 'values', [start, end]);
+  }
+
+  /**
+   * Adds a filter only if it is unique
+   * @param {FilterFragment} filterToAdd - the new filter to add
+   */
+  _doAddFilter(filterToAdd) {
+    const filterExists = this.filters.find(filter => isEqual(filter.serialize(), filterToAdd.serialize()));
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/request.js:280]: Why do we need this check?

File: packages/core/addon/models/bard-request-v2/request.js [hunk 7]
```diff
@@ -110,4 +119,280 @@ export default class Request extends Fragment.extend(Validations) {
...
+    const sort = this.sorts.find(sort => sort.type === 'metric' && sort.canonicalName === metricName);
+    return this.removeSort(sort);
+  }
+
+  /**
+   * Removes a sort if it exists by the given metric with the given parameters
+   * @param {ColumnMetadata} metricMetadataModel - the   metadata of the metric to remove sorts
+   * @param {object} parameters - the parameters applied to the metric
+   */
+  removeSortWithParams(metricMetadataModel, parameters) {
```
Comment (jkusa) [inline packages/core/addon/models/bard-request-v2/request.js:380]: Do we need all these ways to remove a sort?

File: packages/reports/addon/consumers/request/column.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+      request.addColumnFromMeta(columnMetadataModel);
+    },
+
+    /**
+     * @action ADD_COLUMN_WITH_PARAMS
+     * @param route - route that has a model that contains a request property
+     * @param columnMetadataModel - metadata model to add
+     * @param parameters - parameters applied to the column
+     */
+    [RequestActions.ADD_COLUMN_WITH_PARAMS]({ currentModel: { request } }, columnMetadataModel, parameters) {
```
Comment (jkusa) [inline packages/reports/addon/consumers/request/column.js:32]: Do we need both `ADD_COLUMN` & `ADD_COLUMN_WITH_PARAMS`?

File: packages/core/addon/models/bard-request-v2/request.js (hunk 2 above)
Comment (cythrawll) [reply packages/core/addon/models/bard-request-v2/request.js:128]: We have time-dimensions that aren't timegrain columns.

File: packages/core/addon/models/bard-request-v2/fragments/base.js (hunk 1 above)
Comment (kevinhinterlong) [reply packages/core/addon/models/bard-request-v2/fragments/base.js:64]: Yes this should be renamed, added a TODO

**Pull request merged**